src/ray/object_manager/plasma/stats_collector.cc

```diff
@@ -0,0 +1,187 @@
+    num_objects_created_by_worker_++;
+    num_bytes_created_by_worker_ += kDataSize;
+  } else if (kSource == plasma::flatbuf::ObjectSource::RestoredFromStorage) {
+    num_objects_restored_++;
+    num_bytes_restored_ += kDataSize;
+  } else if (kSource == plasma::flatbuf::ObjectSource::ReceivedFromRemoteRaylet) {
+    num_objects_received_++;
+    num_bytes_received_ += kDataSize;
+  } else if (kSource == plasma::flatbuf::ObjectSource::ErrorStoredByRaylet) {
+    num_objects_errored_++;
+    num_bytes_errored_ += kDataSize;
+  }
+
+  RAY_CHECK(!obj.Sealed());
+  num_objects_unsealed_++;
+  num_bytes_unsealed_ += kDataSize;
+}
+
+void ObjectStatsCollector::OnObjectSealed(const LocalObject &obj) {
+  RAY_CHECK(obj.Sealed());
+  const auto kDataSize = obj.GetObjectInfo().data_size;
+
+  num_objects_unsealed_--;
+  num_bytes_unsealed_ -= kDataSize;
+
+  if (obj.GetRefCount() == 1) {
+    if (obj.GetSource() == plasma::flatbuf::ObjectSource::CreatedByWorker) {
+      num_objects_spillable_++;
+      num_bytes_spillable_ += kDataSize;
+    }
+  }
+
+  // though this won't happen in practice but add it here for completeness.
+  if (obj.GetRefCount() == 0) {
+    num_objects_evictable_++;
+    num_bytes_evictable_ += kDataSize;
+  }
+}
+
+void ObjectStatsCollector::OnObjectDeleting(const LocalObject &obj) {
+  const auto kDataSize = obj.GetObjectInfo().data_size;
+  const auto kSource = obj.GetSource();
+
+  if (kSource == plasma::flatbuf::ObjectSource::CreatedByWorker) {
+    num_objects_created_by_worker_--;
+    num_bytes_created_by_worker_ -= kDataSize;
+  } else if (kSource == plasma::flatbuf::ObjectSource::RestoredFromStorage) {
+    num_objects_restored_--;
+    num_bytes_restored_ -= kDataSize;
+  } else if (kSource == plasma::flatbuf::ObjectSource::ReceivedFromRemoteRaylet) {
+    num_objects_received_--;
+    num_bytes_received_ -= kDataSize;
+  } else if (kSource == plasma::flatbuf::ObjectSource::ErrorStoredByRaylet) {
+    num_objects_errored_--;
+    num_bytes_errored_ -= kDataSize;
+  }
+
+  if (obj.GetRefCount() > 0) {
+    num_objects_in_use_--;
+    num_bytes_in_use_ -= kDataSize;
+  }
+
+  if (!obj.Sealed()) {
+    num_objects_unsealed_--;
+    num_bytes_unsealed_ -= kDataSize;
+    return;
+  }
+
+  // obj sealed
+  if (obj.GetRefCount() == 1 &&
+      kSource == plasma::flatbuf::ObjectSource::CreatedByWorker) {
+    num_objects_spillable_--;
+    num_bytes_spillable_ -= kDataSize;
+  }
+
+  if (obj.GetRefCount() == 0) {
+    num_objects_evictable_--;
+    num_bytes_evictable_ -= kDataSize;
+  }
+}
+
+void ObjectStatsCollector::OnObjectRefIncreased(const LocalObject &obj) {
+  const auto kDataSize = obj.GetObjectInfo().data_size;
+  const auto kSource = obj.GetSource();
+  const bool kSealed = obj.Sealed();
+
+  // object ref count bump from 0 to 1
+  if (obj.GetRefCount() == 1) {
+    num_objects_in_use_++;
+    num_bytes_in_use_ += kDataSize;
+
+    if (kSource == plasma::flatbuf::ObjectSource::CreatedByWorker && kSealed) {
```

Comment:
cc @stephanie-wang @rkooo567 this logic is different from PlasmaStore::IsObjectSpillable where here we only treated object created by worker spillable. Should we change this?

Reply:
We currently only can spill primary copies, so it is correct that created by worker source is the only object that’s spillable. (I think created by worker basicaly means the object is primary copy) 

i believe this source is recently added. For IsSpillable, it worked because that’s anyway called only on primary copies. It should also be okay to add this source predicate to IsSpillableObject method (and that will serve the same thing)